src/perf/bin/appmain.cpp

```diff
@@ -0,0 +1,228 @@
+/*++
+
+    Copyright (c) Microsoft Corporation.
+    Licensed under the MIT License.
+
+Abstract:
+
+    QUIC Perf Main execution runner.
+
+--*/
+
+#define QUIC_TEST_APIS 1
+#include "quic_driver_main.h"
+#include "PerfHelpers.h"
+#include <quic_trace.h>
+#ifdef QUIC_CLOG
+#include "appmain.cpp.clog.h"
+#endif
+
+#ifdef _WIN32
+
+//
+// Name of the driver service for quicperf.sys.
+// Must be defined before quic_driver_helpers.h is included
+//
+#define QUIC_DRIVER_NAME   "quicperf"
+
+#include <winioctl.h>
+#include "PerfIoctls.h"
+#include "quic_driver_helpers.h"
+
+#endif
+
+extern "C" _IRQL_requires_max_(PASSIVE_LEVEL) void QuicTraceRundown(void) { }
+
+QUIC_STATUS
+QuicUserMain(
+    _In_ int argc,
+    _In_reads_(argc) _Null_terminated_ char* argv[],
+    _In_ bool KeyboardWait,
+    _In_ PerfSelfSignedConfiguration* SelfSignedConfig
+    ) {
+    QUIC_EVENT StopEvent;
+    QuicEventInitialize(&StopEvent, true, false);
+
+    QUIC_STATUS RetVal = QuicMainStart(argc, argv, StopEvent, SelfSignedConfig);
+    if (RetVal != 0) {
+        return RetVal;
+    }
+
+    printf("Started!\n\n");
+    fflush(stdout);
+
+    if (KeyboardWait) {
+        printf("Press enter to exit\n");
+        getchar();
+        QuicEventSet(StopEvent);
+    }
+
+    RetVal = QuicMainStop(0);
+    QuicEventUninitialize(StopEvent);
+    return RetVal;
+}
+
+#ifdef _WIN32
+
+QUIC_STATUS
+QuicKernelMain(
+    _In_ int argc,
+    _In_reads_(argc) _Null_terminated_ char* argv[],
+    _In_ bool KeyboardWait,
+    _In_ QUIC_SEC_CONFIG_PARAMS* SelfSignedParams
+    ) {
+    size_t TotalLength = 0;
+
+    //
+    // Get total length
+    //
+    for (int i = 0; i < argc; ++i) {
+        //
+        // Length of string, plus null terminator, plus length
+        //
+        TotalLength += strlen(argv[i]) + 1;
+    }
+
+    TotalLength += sizeof(argc);
+
+    char* Data = static_cast<char*>(QUIC_ALLOC_NONPAGED(TotalLength));
+    if (!Data) {
+        printf("Failed to allocate arguments to pass\n");
+        return QUIC_STATUS_OUT_OF_MEMORY;
+    }
+
+    char* DataCurrent = Data;
+
+    QuicCopyMemory(DataCurrent, &argc, sizeof(TotalLength));
+
+    DataCurrent += sizeof(argc);
```

Comment:
```suggestion
    char* DataCurrent = Data;

    *((int*)DataCurrent) = argc;
    DataCurrent += sizeof(argc);
```
Personally, I prefer this approach that a direct call to memcpy. Also, you had a bug above because `sizeof(TotalLength)` does not equal `sizeof(argc)`. This pattern makes that type of bug impossible.

Reply:
That pattern is undefined behavior though. It has the potential for unaligned access, which ARM is especially sensitive about, even on Windows Arm.

Reply:
Why is it any different from accessing a packed struct? This pattern is used all over the Windows code base.

Reply:
Its more the Arm Linux builds I'm worried about. The windows compiler goes out of its way to try and make things like this work. Gcc doesn't do that. In this case its fine, because its the first item added. But if it happened at the end that would cause arm linux to segfault.